### ISSUE TITLE
RoPE and attention mask mismatch

### ISSUE THREAD

**Issue by probablyabot** [ML-GSAI/LLaDA#120]
When doing inference with prompts of different length, having an attention mask is super useful so that model outputs are invariant to padding/batch size. Normally I expect that the model's answer to a question shouldn't change if I add pad tokens or batch it with another question, which is true for normal (autoregressive) LLMs.

For LLaDA, while the authors have recently added support for an attention_mask argument in model.forward(), which correctly masks out any padding tokens from the attention computation, there is still an issue with RoPE that causes the output to depend on batching/padding.

Since the RoPE module uses inferred positions based on only the tensor shape, adding pad tokens on the left (even with a proper attention mask) will shift all the subsequent positions, causing the qk vectors to change.

I think the RoPE module should accept a position_ids argument that allows it to ignore any pad tokens. This along with the attention mask will make the behavior deterministic regardless of padding.

**Comment by nieshenx** [comment]
Thank you very much for your interest in our work!

First, ROPE (Rotary Positional Encoding) is based on relative positional relationships. Therefore, padding on the left or right does not affect ROPE’s effectiveness; what truly matters for ROPE is the relative positions between tokens.

Second, regarding the discrepancy between batch inference results and single-example inference results, we believe this is likely caused by operator-level nondeterminism rather than the model itself. For more details on this issue, please refer to the following blog:
https://thinkingmachines.ai/blog/defeating-nondeterminism-in-llm-inference/

**Comment by probablyabot** [comment]
Theoretically yes, RoPE is designed to operate on relative differences w.r.t. SDPA, however in LLaDA's implementation the actual rotation depends on the absolute position. In other words if I have a sequence [token_a, token_b, ...] and I add a pad token [PAD, token_a, token_b], the qk vectors for token_a are rotated differentlly, and the output changes.

This is easy to verify: try passing the same sequence (batch size 1) to LLaDA with different pad tokens, and you will get different logits. This is different than inference nondeterminism linked in the blog post: if you pass the exact same input twice, you will get the exact same logits.

As for why different rotations lead to different outputs (even though RoPE is designed to be invariant to such shifts), this could be due to numerical precision, or maybe just that LLaDA wasn't trained on sequences where the first token is rotated by a different angle.

But anyway fixing the RoPE module to ignore tokens that are masked out will solve this issue. In the example, this will effectively make token_a the first position in [PAD, token_a, token_b].

**Comment by xushaoxuan123** [comment]
**Thank you for the insight.**
 Following your suggestion, I attempted to verify the fix. Instead of manually slicing the position IDs, I switched the padding side from `left` to `right` (which I believe theoretically achieves the same alignment, ensuring position IDs start from 0).
I also use the first block to simplify the question. The result shows a massive improvement, but there is still a tiny discrepancy: 
>
> *   **Short Stability:** 0.000000
> *   **Long Stability:** 0.000000
> *   **Padding Impact:** 0.032471 (left padding is 0.046875)
>
While `0.03` is likely just floating-point precision noise (BF16/FP16) rather than a semantic error, I am curious about your specific implementation. Could you share your code snippet for handling the `position_ids` with left-padding? I'd like to see if perfect alignment is achievable.

**Comment by probablyabot** [comment]
Thanks for looking into the issue, I think many open source LLMs have a `position_ids` argument. For example, Llama's RoPE (https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L73) has it, although their code is structured a little differently.

I haven't tested the change on LLaDA, but I think it should be easy to ask an LLM to make this change. Here's what Gemini suggested:

```python
def forward(self, q: torch.Tensor, k: torch.Tensor, position_ids: Optional[torch.LongTensor] = None) -> Tuple[torch.Tensor, torch.Tensor]:
        # 1. Prepare the full cache (as you were doing)
        # Ensure we have enough cache for the highest position index requested
        if position_ids is not None:
            max_pos = position_ids.max().item() + 1
        else:
            max_pos = q.shape[-2]
            
        # Get the big table (e.g., size 4096 x 128)
        # This function updates the self.__cache if needed
        full_sin, full_cos = self.get_rotary_embedding(max_pos, q.device)

        # 2. Select the specific rows we need using indexing
        if position_ids is not None:
            # position_ids: [Batch, Seq]
            # full_cos: [1, 1, Max_Len, Dim] -> squeeze to [Max_Len, Dim]
            
            # We use standard PyTorch indexing to "gather" the rows
            # This creates a tensor of shape [Batch, Seq, Dim]
            cos = full_cos.squeeze(0).squeeze(0)[position_ids]
            sin = full_sin.squeeze(0).squeeze(0)[position_ids]
            
            # Unsqueeze to add Head dimension back: [Batch, 1, Seq, Dim]
            cos = cos.unsqueeze(1)
            sin = sin.unsqueeze(1)
        else:
            # Fallback for when no position_ids are provided
            seq_len = q.shape[-2]
            cos = full_cos[:, :, :seq_len, :]
            sin = full_sin[:, :, :seq_len, :]

        # 3. Apply
        q_ = self.apply_rotary_pos_emb(sin, cos, q)
        k_ = self.apply_rotary_pos_emb(sin, cos, k)
        return q_, k_
```
